dpe/src/commands/sign.rs

```diff
@@ -54,28 +56,41 @@ impl<C: Crypto> CommandExecution<C> for SignCmd {
         let digest = Digest::new(&self.digest, DPE_PROFILE.alg_len())
             .map_err(|_| DpeErrorCode::InternalError)?;
         let algs = DPE_PROFILE.alg_len();
-        let priv_key = if self.uses_nd_derivation() {
-            dpe.contexts[idx].cached_priv_key.take().unwrap_or_else({
-                || {
-                    C::derive_ecdsa_key(
-                        algs,
-                        &dpe.derive_cdi(idx, true).unwrap(),
-                        &self.label,
-                        b"ECC",
-                    )
-                }
-            })
-        } else {
+
+        let EcdsaSig { r, s } = if self.uses_symmetric() {
```

Comment:
This block is starting to get pretty complex. Do you think it would be helpful to add a couple local helper functions? They could take care of the key handling internally. It could turn this block into:

```Rust
        let EcdsaSig { r, s } = if !self.uses_symmetric() {
            self.ecdsa_sign(dpe, idx, &digest)?
        } else {
            let r = self.hmac_sign(dpe, idx, &digest)?;
            let s = CryptoBuf::default(DPE_PROFILE.alg_len());
            EcdsaSig { r, s }
        };
```

with these helper functions:

```Rust
    fn ecdsa_sign<C: Crypto>(
        &self,
        dpe: &mut DpeInstance<C>,
        idx: usize,
        digest: &Digest,
    ) -> Result<EcdsaSig, DpeErrorCode> {
        let algs = DPE_PROFILE.alg_len();
        let priv_key = if self.uses_nd_derivation() {
            dpe.contexts[idx].cached_priv_key.take().unwrap_or_else({
                || {
                    C::derive_private_key(
                        algs,
                        &dpe.derive_cdi(idx, true).unwrap(),
                        &self.label,
                        b"ECC",
                    )
                }
            })
        } else {
            let cdi = dpe.derive_cdi(idx, false)?;
            C::derive_private_key(algs, &cdi, &self.label, b"ECC")
        };

        let sig = C::ecdsa_sign_with_derived(algs, digest, &priv_key)
            .map_err(|_| DpeErrorCode::InternalError)?;

        // cache private key
        if self.uses_nd_derivation() {
            dpe.contexts[idx].cached_priv_key.replace(priv_key);
        };

        Ok(sig)
    }

    fn hmac_sign<C: Crypto>(
        &self,
        dpe: &mut DpeInstance<C>,
        idx: usize,
        digest: &Digest,
    ) -> Result<HmacSig, DpeErrorCode> {
        let algs = DPE_PROFILE.alg_len();
        let cdi = dpe.derive_cdi(idx, false)?;
        C::symmetric_sign_with_derived(algs, &cdi, &self.label, b"HMAC", digest)
            .map_err(|_| DpeErrorCode::InternalError)
    }
```